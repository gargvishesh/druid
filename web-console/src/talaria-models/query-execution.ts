/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { QueryResult, SqlExpression, SqlQuery, SqlWithQuery } from 'druid-query-toolkit';

import { deepGet, oneOf } from '../utils';
import { QueryContext } from '../utils/query-context';

import { StageDefinition } from './talaria-stage';

// Hack around the concept that we might get back a SqlWithQuery and will need to unpack it
function parseSqlQuery(queryString: string): SqlQuery | undefined {
  const q = SqlExpression.maybeParse(queryString);
  if (!q) return;
  if (q instanceof SqlWithQuery) return q.flattenWith();
  if (q instanceof SqlQuery) return q;
  return;
}

export interface TalariaTaskError {
  error: {
    errorCode: string;
    errorMessage?: string;
    [key: string]: any;
  };
  host?: string;
  taskId?: string;
  stageNumber?: number;
  exceptionStackTrace?: string;
}

type ExecutionDestination =
  | {
      type: 'taskReport';
    }
  | { type: 'dataSource'; dataSource: string; exists?: boolean }
  | { type: 'external' }
  | { type: 'download' };

export type QueryExecutionStatus = 'RUNNING' | 'FAILED' | 'SUCCESS';

export interface QueryExecutionValue {
  id: string;
  sqlQuery?: string;
  queryContext?: QueryContext;
  status?: QueryExecutionStatus;
  startTime?: Date;
  duration?: number;
  stages?: StageDefinition[];
  destination?: ExecutionDestination;
  result?: QueryResult;
  error?: TalariaTaskError;
}

export class QueryExecution {
  static validStatus(
    status: string | undefined,
  ): status is 'WAITING' | 'PENDING' | 'RUNNING' | 'FAILED' | 'SUCCESS' {
    return oneOf(status, 'WAITING', 'PENDING', 'RUNNING', 'FAILED', 'SUCCESS');
  }

  static normalizeAsyncStatus(
    state: 'INITIALIZED' | 'RUNNING' | 'COMPLETE' | 'FAILED' | 'UNDETERMINED',
  ): QueryExecutionStatus {
    switch (state) {
      case 'COMPLETE':
        return 'SUCCESS';

      case 'INITIALIZED':
      case 'UNDETERMINED':
        return 'RUNNING';

      default:
        return state;
    }
  }

  // Treat WAITING as PENDING since they are all the same as far as the UI is concerned
  static normalizeTaskStatus(
    status: 'WAITING' | 'PENDING' | 'RUNNING' | 'FAILED' | 'SUCCESS',
  ): QueryExecutionStatus {
    switch (status) {
      case 'SUCCESS':
      case 'FAILED':
        return status;

      default:
        return 'RUNNING';
    }
  }

  static fromAsyncStatus(
    asyncResult: any,
    sqlQuery?: string,
    queryContext?: QueryContext,
  ): QueryExecution {
    const status = QueryExecution.normalizeAsyncStatus(asyncResult.state);
    return new QueryExecution({
      id: asyncResult.asyncResultId,
      status: asyncResult.error ? 'FAILED' : status,
      sqlQuery,
      queryContext,
      error: asyncResult.error
        ? { error: { errorCode: 'AsyncError', errorMessage: JSON.stringify(asyncResult.error) } }
        : status === 'FAILED'
        ? {
            error: {
              errorCode: 'UnknownError',
              errorMessage:
                'Execution failed, there is no detail information, and there is no error in the status response',
            },
          }
        : undefined,
      destination:
        asyncResult.engine === 'Broker'
          ? {
              type: 'taskReport',
            }
          : undefined,
    });
  }

  static fromAsyncDetail(report: any): QueryExecution {
    // Must have status set for a valid report
    const id = deepGet(report, 'talariaStatus.taskId') || deepGet(report, 'talariaTask.payload.id');

    let status = deepGet(report, 'talariaStatus.payload.status');
    if (
      !status &&
      report.error &&
      !String(report.error).startsWith(`Can't find chatHandler for handler`) // Ignore this error in particular
    ) {
      status = 'FAILED';
    }

    if (typeof id !== 'string' || !QueryExecution.validStatus(status)) {
      throw new Error('Invalid payload');
    }

    let error: TalariaTaskError | undefined;
    if (status === 'FAILED') {
      error =
        deepGet(report, 'talariaStatus.payload.errorReport') ||
        (typeof report.error === 'string'
          ? { error: { errorCode: 'UnknownError', errorMessage: report.error } }
          : undefined);
    }

    const stages = deepGet(report, 'talariaStages.payload.stages');
    const startTime = new Date(deepGet(report, 'talariaStatus.payload.startTime'));
    const durationMs = deepGet(report, 'talariaStatus.payload.durationMs');
    let res = new QueryExecution({
      id,
      status: QueryExecution.normalizeTaskStatus(status),
      startTime: isNaN(startTime.getTime()) ? undefined : startTime,
      duration: typeof durationMs === 'number' ? durationMs : undefined,
      stages,
      error,
      destination: deepGet(report, 'talariaTask.payload.spec.destination'),
    });

    if (deepGet(report, 'talariaTask.payload.sqlQuery')) {
      res = res.changeSqlQuery(
        deepGet(report, 'talariaTask.payload.sqlQuery'),
        deepGet(report, 'talariaTask.payload.sqlQueryContext'),
      );
    }

    return res;
  }

  static fromResult(result: QueryResult): QueryExecution {
    return new QueryExecution({
      id: result.sqlQueryId || result.queryId || 'direct_result',
      status: 'SUCCESS',
      result,
    });
  }

  public readonly id: string;
  public readonly sqlQuery?: string;
  public readonly queryContext?: QueryContext;
  public readonly status?: QueryExecutionStatus;
  public readonly startTime?: Date;
  public readonly duration?: number;
  public readonly stages?: StageDefinition[];
  public readonly destination?: ExecutionDestination;
  public readonly result?: QueryResult;
  public readonly error?: TalariaTaskError;

  constructor(value: QueryExecutionValue) {
    this.id = value.id;
    this.sqlQuery = value.sqlQuery;
    this.queryContext = value.queryContext;
    this.status = value.status;
    this.startTime = value.startTime;
    this.duration = value.duration;
    this.stages = value.stages;
    this.destination = value.destination;
    this.result = value.result;
    this.error = value.error;
  }

  valueOf(): QueryExecutionValue {
    return {
      id: this.id,
      sqlQuery: this.sqlQuery,
      queryContext: this.queryContext,
      status: this.status,
      startTime: this.startTime,
      duration: this.duration,
      stages: this.stages,
      destination: this.destination,
      result: this.result,
      error: this.error,
    };
  }

  public changeSqlQuery(sqlQuery: string, queryContext?: QueryContext): QueryExecution {
    const value = this.valueOf();

    value.sqlQuery = sqlQuery;
    value.queryContext = queryContext;
    const parsedQuery = parseSqlQuery(sqlQuery);
    if (value.result && (parsedQuery || queryContext)) {
      value.result = value.result.attachQuery({ context: queryContext }, parsedQuery);
    }

    return new QueryExecution(value);
  }

  public changeDestination(destination: ExecutionDestination): QueryExecution {
    return new QueryExecution({
      ...this.valueOf(),
      destination,
    });
  }

  public changeResult(result: QueryResult): QueryExecution {
    return new QueryExecution({
      ...this.valueOf(),
      result: result.attachQuery({}, this.sqlQuery ? parseSqlQuery(this.sqlQuery) : undefined),
    });
  }

  public updateWith(newSummary: QueryExecution): QueryExecution {
    let nextSummary = newSummary;
    if (this.sqlQuery && !nextSummary.sqlQuery) {
      nextSummary = nextSummary.changeSqlQuery(this.sqlQuery, this.queryContext);
    }
    if (this.destination && !nextSummary.destination) {
      nextSummary = nextSummary.changeDestination(this.destination);
    }

    return nextSummary;
  }

  public attachErrorFromStatus(status: any): QueryExecution {
    const errorMsg = deepGet(status, 'status.errorMsg');

    return new QueryExecution({
      ...this.valueOf(),
      error: {
        error: {
          errorCode: 'UnknownError',
          errorMessage: errorMsg,
        },
      },
    });
  }

  public markDestinationDatasourceExists(): QueryExecution {
    const { destination } = this;
    if (destination?.type !== 'dataSource') return this;

    return new QueryExecution({
      ...this.valueOf(),
      destination: {
        ...destination,
        exists: true,
      },
    });
  }

  public isWaitingForQuery(): boolean {
    const { status } = this;
    return status !== 'SUCCESS' && status !== 'FAILED';
  }

  public isFullyComplete(): boolean {
    if (this.isWaitingForQuery()) return false;

    const { status, destination } = this;
    if (status === 'SUCCESS' && destination?.type === 'dataSource') {
      return Boolean(destination.exists);
    }

    return true;
  }

  public getInsertDatasource(): string | undefined {
    const { destination } = this;
    if (destination?.type !== 'dataSource') return;
    return destination.dataSource;
  }

  public isSuccessfulInsert(): boolean {
    return Boolean(
      this.isFullyComplete() && this.getInsertDatasource() && this.status === 'SUCCESS',
    );
  }

  public getErrorMessage(): string | undefined {
    const { error } = this;
    if (!error) return;
    return (
      (error.error.errorCode ? `${error.error.errorCode}: ` : '') +
      (error.error.errorMessage || (error.exceptionStackTrace || '').split('\n')[0])
    );
  }

  public getEndTime(): Date | undefined {
    const { startTime, duration } = this;
    if (!startTime || !duration) return;
    return new Date(startTime.valueOf() + duration);
  }
}
