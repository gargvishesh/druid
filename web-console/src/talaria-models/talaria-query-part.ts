/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { SqlExpression, SqlQuery, SqlRef, SqlTableRef, SqlWithQuery } from 'druid-query-toolkit';
import Hjson from 'hjson';
import * as JSONBig from 'json-bigint-native';

import { ColumnMetadata, compact, filterMap } from '../utils';

import { generate8HexId } from './talaria-general';
import { fitExternalConfigPattern } from './talaria-query-pattern';

function sqlTypeFromDruid(druidType: string): string {
  druidType = druidType.toLowerCase();
  switch (druidType) {
    case 'string':
      return 'VARCHAR';

    case 'long':
      return 'BIGINT';

    case 'float':
    case 'double':
      return druidType.toUpperCase();

    default:
      return 'COMPLEX';
  }
}

// -----------------------------

export interface TalariaQueryPartValue {
  id: string;
  queryName?: string;
  queryString: string;
  collapsed?: boolean;
  lastQueryId?: string;
}

export class TalariaQueryPart {
  static blank() {
    return new TalariaQueryPart({
      id: generate8HexId(),
      queryString: '',
    });
  }

  static fromQuery(query: SqlQuery, queryName?: string, collapsed?: boolean) {
    return this.fromQueryString(query.changeParens([]).toString(), queryName, collapsed);
  }

  static fromQueryString(queryString: string, queryName?: string, collapsed?: boolean) {
    return new TalariaQueryPart({
      id: generate8HexId(),
      queryName,
      queryString,
      collapsed,
    });
  }

  static isTalariaEngineNeeded(queryString: string): boolean {
    return /EXTERN\s*\(/im.test(queryString) || /INSERT\s+INTO/im.test(queryString);
  }

  public readonly id: string;
  public readonly queryName?: string;
  public readonly queryString: string;
  public readonly collapsed: boolean;
  public readonly lastQueryId?: string;

  public readonly parsedQuery?: SqlQuery;

  constructor(value: TalariaQueryPartValue) {
    this.id = value.id;
    this.queryName = value.queryName;
    this.queryString = value.queryString;
    this.collapsed = Boolean(value.collapsed);
    this.lastQueryId = value.lastQueryId;

    try {
      this.parsedQuery = SqlQuery.parse(this.queryString);
    } catch {}
  }

  public valueOf(): TalariaQueryPartValue {
    return {
      id: this.id,
      queryName: this.queryName,
      queryString: this.queryString,
      collapsed: this.collapsed,
      lastQueryId: this.lastQueryId,
    };
  }

  public changeId(id: string): TalariaQueryPart {
    return new TalariaQueryPart({ ...this.valueOf(), id });
  }

  public changeQueryName(queryName: string): TalariaQueryPart {
    return new TalariaQueryPart({ ...this.valueOf(), queryName });
  }

  public changeQueryString(queryString: string): TalariaQueryPart {
    return new TalariaQueryPart({ ...this.valueOf(), queryString });
  }

  public changeCollapsed(collapsed: boolean): TalariaQueryPart {
    return new TalariaQueryPart({ ...this.valueOf(), collapsed });
  }

  public changeLastQueryId(lastQueryId: string | undefined): TalariaQueryPart {
    return new TalariaQueryPart({ ...this.valueOf(), lastQueryId });
  }

  public clear(): TalariaQueryPart {
    return new TalariaQueryPart({
      ...this.valueOf(),
      queryString: '',
    });
  }

  public isEmptyQuery(): boolean {
    return this.queryString.trim() === '';
  }

  public isJsonLike(): boolean {
    return this.queryString.trim().startsWith('{');
  }

  public validRune(): boolean {
    try {
      Hjson.parse(this.queryString);
      return true;
    } catch {
      return false;
    }
  }

  public prettyPrintJson(): TalariaQueryPart {
    let parsed: unknown;
    try {
      parsed = Hjson.parse(this.queryString);
    } catch {
      return this;
    }
    return this.changeQueryString(JSONBig.stringify(parsed, undefined, 2));
  }

  public getInsertDatasource(): string | undefined {
    const { queryString, parsedQuery } = this;
    if (parsedQuery) {
      return parsedQuery.getInsertIntoTable()?.getTable();
    }

    if (this.isJsonLike()) return;

    // When the parser fails try a regexp
    const m = /INSERT\s+INTO\s*([^\n]+)/im.exec(queryString);
    if (m) {
      try {
        const ex = SqlExpression.parse(m[1]);
        if (ex instanceof SqlRef) {
          return ex.getColumn();
        }
      } catch {}
    }

    return;
  }

  public getInlineMetadata(): ColumnMetadata[] {
    const { queryName, parsedQuery } = this;
    if (queryName && parsedQuery) {
      try {
        return fitExternalConfigPattern(parsedQuery).columns.map(({ name, type }) => ({
          COLUMN_NAME: name,
          DATA_TYPE: sqlTypeFromDruid(type),
          TABLE_NAME: queryName,
          TABLE_SCHEMA: 'druid',
        }));
      } catch {
        return filterMap(parsedQuery.getSelectExpressionsArray(), ex => {
          const outputName = ex.getOutputName();
          if (!outputName) return;
          return {
            COLUMN_NAME: outputName,
            DATA_TYPE: 'UNKNOWN',
            TABLE_NAME: queryName,
            TABLE_SCHEMA: 'druid',
          };
        });
      }
    }
    return [];
  }

  public isTalariaEngineNeeded(): boolean {
    return TalariaQueryPart.isTalariaEngineNeeded(this.queryString);
  }

  public explodeQueryPart(): TalariaQueryPart[] {
    let flatQuery: SqlQuery;
    try {
      // We need to do our own parsing here because this.parseQuery necessarily must be a SqlQuery
      // object, and we might have a SqlWithQuery here.
      flatQuery = (SqlExpression.parse(this.queryString) as SqlWithQuery).flattenWith();
    } catch {
      return [this];
    }

    const possibleNewParts = flatQuery.getWithParts().map(({ table, columns, query }) => {
      if (columns) return;
      return TalariaQueryPart.fromQuery(query, table.name, true);
    });

    const newParts = compact(possibleNewParts);
    if (newParts.length !== possibleNewParts.length) {
      return [this];
    }

    return newParts.concat(this.changeQueryString(flatQuery.changeWithParts(undefined).toString()));
  }

  public toWithPart(): string {
    const { queryName, queryString } = this;
    return `${SqlTableRef.create(queryName || 'q')} AS (\n${queryString}\n)`;
  }

  public duplicate(): TalariaQueryPart {
    return this.changeId(generate8HexId()).changeLastQueryId(undefined);
  }
}
