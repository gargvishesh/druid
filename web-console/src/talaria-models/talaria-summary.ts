/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { Column, QueryResult, SqlQuery } from 'druid-query-toolkit';

import { deepGet, oneOf } from '../utils';
import { QueryContext } from '../utils/query-context';

import { getStartTime, StageDefinition } from './talaria-stage';

interface TalariaResultsPayload {
  results: any[][];
  signature: { name: string; type: string }[];
  sqlTypeNames?: string[];
}

export interface TalariaTaskError {
  error: {
    errorCode: string;
    errorMessage?: string;
    [key: string]: any;
  };
  host?: string;
  taskId?: string;
  stageNumber?: number;
  exceptionStackTrace?: string;
}

type TalariaDestination =
  | {
      type: 'taskReport';
    }
  | { type: 'external' }
  | { type: 'dataSource'; dataSource: string; exists?: boolean };

export type TalariaSummarySource = 'init' | 'reattach' | 'report' | 'status';
export type TalariaSummaryStatus = 'PENDING' | 'RUNNING' | 'FAILED' | 'SUCCESS';

export interface TalariaSummaryValue {
  taskId: string;
  summarySource: TalariaSummarySource;
  sqlQuery?: string;
  queryContext?: QueryContext;
  status?: TalariaSummaryStatus;
  startTime?: Date;
  stages?: StageDefinition[];
  destination?: TalariaDestination;
  result?: QueryResult;
  error?: TalariaTaskError;
}

export class TalariaSummary {
  static validStatus(
    status: string | undefined,
  ): status is 'WAITING' | 'PENDING' | 'RUNNING' | 'FAILED' | 'SUCCESS' {
    return oneOf(status, 'WAITING', 'PENDING', 'RUNNING', 'FAILED', 'SUCCESS');
  }

  // Treat WAITING as PENDING since they are all the same as far as the UI is concerned
  static normalizeStatus(
    status: 'WAITING' | 'PENDING' | 'RUNNING' | 'FAILED' | 'SUCCESS',
  ): 'PENDING' | 'RUNNING' | 'FAILED' | 'SUCCESS' {
    return status === 'WAITING' ? 'PENDING' : status;
  }

  static resultsPayloadToQueryResult(resultsPayload: TalariaResultsPayload, taskId: string) {
    const { results, signature, sqlTypeNames } = resultsPayload;
    return new QueryResult({
      header: signature.map(
        (sig, i) =>
          new Column({ name: sig.name, nativeType: sig.type, sqlType: sqlTypeNames?.[i] }),
      ),
      rows: results,
    })
      .inflateDatesFromSqlTypes()
      .changeResultContext({ taskId });
    // .changeQueryDuration(...) // ToDo: add query duration one day
  }

  static init(taskId: string, sqlQuery?: string, queryContext?: QueryContext): TalariaSummary {
    return new TalariaSummary({
      taskId,
      summarySource: 'init',
      sqlQuery,
      queryContext,
    });
  }

  static reattach(taskId: string): TalariaSummary {
    return new TalariaSummary({
      taskId,
      summarySource: 'reattach',
    });
  }

  static fromReport(report: any): TalariaSummary {
    // Must have status set for a valid report
    const taskId = deepGet(report, 'talariaStatus.taskId');
    const status = deepGet(report, 'talariaStatus.payload.status');
    if (typeof taskId !== 'string' || !TalariaSummary.validStatus(status)) {
      throw new Error('invalid status');
    }

    let error: TalariaTaskError | undefined;
    if (status === 'FAILED') {
      error = deepGet(report, 'talariaStatus.payload.errorReport');
    }

    const stages = deepGet(report, 'talariaStages.payload.stages') || [];
    const results = deepGet(report, 'talariaResults.payload');
    return new TalariaSummary({
      taskId,
      summarySource: 'report',
      status: TalariaSummary.normalizeStatus(status),
      startTime: getStartTime(stages),
      stages,
      destination: results ? { type: 'taskReport' } : undefined,
      result: results ? TalariaSummary.resultsPayloadToQueryResult(results, taskId) : undefined,
      error,
    });
  }

  static fromStatus(statusPayload: any): TalariaSummary {
    const taskId = deepGet(statusPayload, 'task');
    const status = deepGet(statusPayload, 'status.status');
    if (typeof taskId !== 'string' || !TalariaSummary.validStatus(status)) {
      throw new Error(`invalid task status from status endpoint: ${status}`);
    }

    let error: TalariaTaskError | undefined;
    if (status === 'FAILED') {
      const errorMsg = deepGet(statusPayload, 'status.errorMsg');
      if (errorMsg) {
        error = {
          error: { errorCode: 'UnknownError', errorMessage: errorMsg },
        };
      }
    }

    let startTime: Date | undefined = new Date(deepGet(statusPayload, 'createdTime'));
    if (isNaN(startTime.valueOf())) startTime = undefined;

    return new TalariaSummary({
      taskId,
      summarySource: 'status',
      status: status === 'WAITING' ? 'PENDING' : status, // Treat WAITING as PENDING since they are all the same for the UI
      startTime,
      error,
    });
  }

  static fromResult(result: QueryResult): TalariaSummary {
    return new TalariaSummary({
      taskId: result.sqlQueryId || result.queryId || 'direct_result',
      summarySource: 'init',
      status: 'SUCCESS',
      result,
    });
  }

  public readonly taskId: string;
  public readonly summarySource: TalariaSummarySource;
  public readonly sqlQuery?: string;
  public readonly queryContext?: QueryContext;
  public readonly status?: TalariaSummaryStatus;
  public readonly startTime?: Date;
  public readonly stages?: StageDefinition[];
  public readonly destination?: TalariaDestination;
  public readonly result?: QueryResult;
  public readonly error?: TalariaTaskError;

  constructor(value: TalariaSummaryValue) {
    this.taskId = value.taskId;
    this.summarySource = value.summarySource;
    this.sqlQuery = value.sqlQuery;
    this.queryContext = value.queryContext;
    this.status = value.status;
    this.startTime = value.startTime;
    this.stages = value.stages;
    this.destination = value.destination;
    this.result = value.result;
    this.error = value.error;
  }

  valueOf(): TalariaSummaryValue {
    return {
      taskId: this.taskId,
      summarySource: this.summarySource,
      sqlQuery: this.sqlQuery,
      queryContext: this.queryContext,
      status: this.status,
      startTime: this.startTime,
      stages: this.stages,
      destination: this.destination,
      result: this.result,
      error: this.error,
    };
  }

  public changeSqlQuery(sqlQuery: string, queryContext?: QueryContext): TalariaSummary {
    const value = this.valueOf();

    value.sqlQuery = sqlQuery;
    value.queryContext = queryContext;
    const parsedQuery = SqlQuery.maybeParse(sqlQuery);
    if (value.result && (parsedQuery || queryContext)) {
      value.result = value.result.attachQuery({ context: queryContext }, parsedQuery);
    }

    return new TalariaSummary(value);
  }

  public changeDestination(destination: TalariaDestination): TalariaSummary {
    return new TalariaSummary({
      ...this.valueOf(),
      destination,
    });
  }

  public updateWith(newSummary: TalariaSummary): TalariaSummary {
    if (this.stages && !newSummary.stages) return this;

    let nextSummary = newSummary;
    if (this.sqlQuery && !nextSummary.sqlQuery) {
      nextSummary = nextSummary.changeSqlQuery(this.sqlQuery, this.queryContext);
    }

    return nextSummary;
  }

  public attachErrorFromStatus(status: any): TalariaSummary {
    const errorMsg = deepGet(status, 'status.errorMsg');

    return new TalariaSummary({
      ...this.valueOf(),
      error: {
        error: {
          errorCode: 'UnknownError',
          errorMessage: errorMsg,
        },
      },
    });
  }

  public needsInfoFromPayload(): boolean {
    return Boolean(!this.destination || !this.sqlQuery);
  }

  public updateFromPayload(payload: any): TalariaSummary {
    return this.changeDestination(deepGet(payload, 'payload.spec.destination')).changeSqlQuery(
      deepGet(payload, 'payload.sqlQuery'),
      deepGet(payload, 'payload.sqlQueryContext'),
    );
  }

  public markDestinationDatasourceExists(): TalariaSummary {
    const { destination } = this;
    if (destination?.type !== 'dataSource') return this;

    return new TalariaSummary({
      ...this.valueOf(),
      destination: {
        ...destination,
        exists: true,
      },
    });
  }

  public isReattach(): boolean {
    return this.summarySource === 'reattach';
  }

  public isWaitingForTask(): boolean {
    const { status } = this;
    return status !== 'SUCCESS' && status !== 'FAILED';
  }

  public isFullyComplete(): boolean {
    if (this.isWaitingForTask()) return false;

    const { status, destination } = this;
    if (status === 'SUCCESS' && destination?.type === 'dataSource') {
      return Boolean(destination.exists);
    }

    return true;
  }

  public getDuration(): number {
    return 0;
  }

  public getInsertDataSource(): string | undefined {
    const { destination } = this;
    if (destination?.type !== 'dataSource') return;
    return destination.dataSource;
  }

  public isSuccessfulInsert(): boolean {
    return Boolean(
      this.isFullyComplete() && this.getInsertDataSource() && this.status === 'SUCCESS',
    );
  }

  public getErrorMessage(): string | undefined {
    const { error } = this;
    if (!error) return;
    return (
      (error.error.errorCode ? `${error.error.errorCode}: ` : '') +
      (error.error.errorMessage || (error.exceptionStackTrace || '').split('\n')[0])
    );
  }
}
