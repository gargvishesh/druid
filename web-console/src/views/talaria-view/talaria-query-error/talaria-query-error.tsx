/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { Intent } from '@blueprintjs/core';
import copy from 'copy-to-clipboard';
import React, { useState } from 'react';

import { ShowValueDialog } from '../../../dialogs/show-value-dialog/show-value-dialog';
import { AppToaster } from '../../../singletons';
import { TalariaTaskError } from '../../../talaria-models';

import './talaria-query-error.scss';

const ClickToCopy = React.memo(function ClickToCopy(props: { text: string }) {
  const { text } = props;
  return (
    <span
      className="click-to-copy"
      title={`Click to copy:\n${text}`}
      onClick={() => {
        copy(text, { format: 'text/plain' });
        AppToaster.show({
          message: `${text} copied to clipboard`,
          intent: Intent.SUCCESS,
        });
      }}
    >
      {text}
    </span>
  );
});

export interface TalariaQueryErrorProps {
  taskError: TalariaTaskError;
}

export const TalariaQueryError = React.memo(function TalariaQueryError(
  props: TalariaQueryErrorProps,
) {
  const { taskError } = props;
  const [stackToShow, setStackToShow] = useState<string | undefined>();

  const { error, exceptionStackTrace, taskId, host } = taskError;

  return (
    <div className="talaria-query-error">
      <p>
        {error.errorCode && <>{`${error.errorCode}: `}</>}
        {error.errorMessage || (exceptionStackTrace || '').split('\n')[0]}
        {exceptionStackTrace && (
          <>
            {' '}
            <span
              className="stack-trace"
              onClick={() => {
                setStackToShow(exceptionStackTrace);
              }}
            >
              (Stack trace)
            </span>
          </>
        )}
      </p>
      {taskId && (
        <p>
          Failed task ID: <ClickToCopy text={taskId} />
        </p>
      )}
      {host && (
        <p>
          On host: <ClickToCopy text={host} />
        </p>
      )}
      {stackToShow && (
        <ShowValueDialog size="large" onClose={() => setStackToShow(undefined)} str={stackToShow} />
      )}
    </div>
  );
});
