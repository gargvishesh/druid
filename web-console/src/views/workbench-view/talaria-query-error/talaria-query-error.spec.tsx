/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { shallow } from 'enzyme';
import React from 'react';

import { Execution } from '../../../talaria-models';

import { TalariaQueryError } from './talaria-query-error';

describe('TalariaQueryError', () => {
  it('matches snapshot', () => {
    const queryError = shallow(
      <TalariaQueryError
        execution={
          new Execution({
            engine: 'sql-task',
            id: 'talaria1_kttm',
            error: {
              taskId: 'talaria1_kttm_naobfgjd_2021-12-03T23:16:15.269Z',
              host: 'ip-10-3-228-20.ec2.internal:8100',
              stageNumber: 0,
              error: {
                errorCode: 'TooManyBuckets',
                maxBuckets: 10000,
                errorMessage:
                  'Too many partition buckets (max = 10,000); try breaking your query up into smaller queries or using a wider segmentGranularity',
              },
              exceptionStackTrace:
                'io.imply.druid.talaria.frame.cluster.statistics.TooManyBucketsException: Too many buckets; maximum is [10000]\n\tat io.imply.druid.talaria.frame.cluster.statistics.ClusterByStatisticsCollectorImpl.getOrCreateBucketHolder(ClusterByStatisticsCollectorImpl.java:324)\n\tat io.imply.druid.talaria.frame.cluster.statistics.ClusterByStatisticsCollectorImpl.add(ClusterByStatisticsCollectorImpl.java:105)\n\tat io.imply.druid.talaria.indexing.TalariaClusterByStatisticsCollectorWorker.runIncrementally(TalariaClusterByStatisticsCollectorWorker.java:87)\n\tat io.imply.druid.talaria.frame.processor.FrameProcessors$1FrameProcessorWithBaggage.runIncrementally(FrameProcessors.java:136)\n\tat io.imply.druid.talaria.frame.processor.FrameProcessorExecutor$1ExecutorRunnable.runProcessorNow(FrameProcessorExecutor.java:197)\n\tat io.imply.druid.talaria.frame.processor.FrameProcessorExecutor$1ExecutorRunnable.run(FrameProcessorExecutor.java:112)\n\tat io.imply.druid.talaria.indexing.TalariaWorkerTask$2$2.run(TalariaWorkerTask.java:731)\n\tat java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)\n\tat java.util.concurrent.FutureTask.run(FutureTask.java:266)\n\tat org.apache.druid.query.PrioritizedListenableFutureTask.run(PrioritizedExecutorService.java:251)\n\tat java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)\n\tat java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)\n\tat java.lang.Thread.run(Thread.java:748)\n',
            },
          })
        }
      />,
    );

    expect(queryError).toMatchSnapshot();
  });
});
