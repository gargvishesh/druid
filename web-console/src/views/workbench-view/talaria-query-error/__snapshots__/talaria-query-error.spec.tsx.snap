// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TalariaQueryError matches snapshot 1`] = `
<div
  className="talaria-query-error"
>
  <p
    className="error-message-text"
  >
    TooManyBuckets: 
    Too many partition buckets (max = 10,000); try breaking your query up into smaller queries or using a wider segmentGranularity
     
    <span
      className="stack-trace"
      onClick={[Function]}
    >
      (Stack trace)
    </span>
  </p>
  <p>
    Failed task ID: 
    <Memo(ClickToCopy)
      text="talaria1_kttm_naobfgjd_2021-12-03T23:16:15.269Z"
    />
     
    On host: 
    <Memo(ClickToCopy)
      text="ip-10-3-228-20.ec2.internal:8100"
    />
     
    Debug:
     
    <span
      className="download-infopack"
      onClick={[Function]}
    >
      download query profile
    </span>
  </p>
</div>
`;
